### DISCUSSION TITLE
Machine learning for spam detection

### DISCUSSION THREAD

**Discussion by Kurdakov** [ThioJoe/YT-Spammer-Purge#298] [Ideas]
while it might be long term target, apparently, spam logs are treasure for possible machine learning applications.

there are quite a bit of attempts on github, so might be reused for training either Naive Bayes/Decision tree  classifiers or even neural network

https://github.com/Savjee/yt-spam-classifier
https://github.com/akhilsali/YouTube-Spam-Filter
https://github.com/Euno257/Youtube-Comments-Spam-Detector
https://github.com/ma-shamshiri/Spam-Detector
https://github.com/amj18/mlwebapp
https://github.com/akshaybahadur21/Spam-Classifier
https://github.com/essamik/spam-classifier-python
https://github.com/subhashbylaiah/spam-classification
https://github.com/ElliotVilhelm/Spam-Detector-LSTM
https://github.com/MGCodesandStats/tensorflow-nlp
https://github.com/chongjason914/spam-classification
etc

so as a first step it would be nice to think about safe possibility to upload logs from YT YouTubeSpammerPurge
maybe just Google Drive and somehow giving access to it to volunteers (or vise versa  let them upload logs somewhere (the same Google Drive and then just collect compressed logs by getting their links here in responses?)

**Comment by KendallDoesCoding** [comment]
> Referen

You mean that any scammers whose comments were deleted will come in the public logs. If no, I would like to see that too, so anyone following this projects, knows those scammers YT Name, channel url/id, etc, if that would be possible

**Comment by majestrate** [comment]
A related idea in the realm of ML powered classification that may be quite useful in relation to evading ramping up the spammer arms race is a ZKANN (zero knowledge adversarial neural network). A spammer would not be able to use the classifier in a GANN as easily in such a case.

**Comment by z0del** [comment]
hello guys. just saw this project on LTT. 
I wonder if the model coud be trained locally since the user already have the labeled data. but so far, it's possible to correct false positives?

**Reply by z0del** [reply]
the current approach using regex could be used to make "weak-labels", then the user corrects the false-positives and now we have the data we can use to train a model..

**Reply by ikkysleepy** [reply]
I tried the naive bayes classifier on a project that is similar to this and got about 97% on a dataset, which sounds good but when I tried a few new comments it got 91%, but not sure what this means. I think the issue is that the model is looking at probability and not patterns and something like this is looking at pattens and not probability. Maybe I am doing something wrong , which very well could be but it looks like pattern detection is better than possibilities detection in this case. Having said that , I will keep trying to do more test with naive bayes, maybe the dataset needs to be bigger.

**Reply by ThioJoe** [reply]
I did a bit of messing around with model training in Azure, but haven't really figured out how to implement it. I added a feature to the program that lets you save the results as JSON, and also save "all comments" scanned too. So after excluding the false positives, anything that was matched will get marked as spam and anything else not spam in the json file, which should make it easy to train assuming false positives are well removed

**Reply by Irteeja** [reply]
what is the ratio of your train and test data?

**Comment by TechStudent10** [comment]
Another (and probably easier) method would be that when the program finds spam, it automatically commits/creates a PR on the [`ThioJoe/YT-Spam-Lists`](https://github.com/ThioJoe/YT-Spam-Lists) repo. That way the program can automatically reference these lists and use them for spam detection.

**Reply by Firecul** [reply]
Discussing in #709
